### PR TITLE
SONARJAVA-958 & SONARJAVA-1179 Migrate Unused private method to semantic analysis

### PR DESCRIPTION
**Pull request by mpaladin**
http://jira.sonarsource.com/browse/SONARJAVA-958
https://jira.sonarsource.com/browse/SONARJAVA-1179

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/java/checks/helpers' package (org.sonarsource.java:java-checks:src/main/java/org/sonar/java/checks/helpers) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: java-checks/src/main/java/org/sonar/java/checks/SerializableContract.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 public final class SerializableContract {
 
-  private static final Set<String> SERIALIZABLE_CONTRACT_METHODS = Sets.newHashSet();
+  public static final Set<String> SERIALIZABLE_CONTRACT_METHODS = Sets.newHashSet();
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SerializableContract.java]: CRITICAL Make this member "protected". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2386)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* CRITICAL 1 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/java/checks/helpers' package (org.sonarsource.java:java-checks:src/main/java/org/sonar/java/checks/helpers) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java [hunk 2]
```diff
@@ -46,47 +40,24 @@
...
-      int line = getMethodLineNumber(asmMethod);
-      if (line > 0) {
-        message.setLine(line);
+  public void visitNode(Tree tree) {
+    MethodTree node = (MethodTree) tree;
+    Symbol symbol = node.symbol();
+    if (symbol.isPrivate() && symbol.usages().isEmpty()) {
+      if (node.is(Tree.Kind.CONSTRUCTOR)) {
+        addIssue(node, "Remove this unused private \"" + node.simpleName().name() + "\" constructor.");
+      } else if (!SerializableContract.SERIALIZABLE_CONTRACT_METHODS.contains(node.symbol().name())) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: you can use `symbol.name()` here.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java [hunk 3]
```diff
@@ -46,47 +40,24 @@
...
-      file.log(message);
     }
   }
 
-  private static boolean isPrivateUnused(AsmMethod asmMethod) {
-    return !asmMethod.isUsed() && asmMethod.isPrivate();
-  }
-
-  private static boolean isExcludedFromCheck(AsmMethod asmMethod) {
-    return asmMethod.isSynthetic() || asmMethod.isDefaultConstructor() || SerializableContract.methodMatch(asmMethod);
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: Previous implementation was ignoring default constructor.

File: java-checks/src/test/files/checks/UnusedPrivateMethod.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    // this method should not be considered as dead code, see Serializable contract
+    return null;
+  }
+
+  private void readObjectNoData() throws java.io.ObjectStreamException {
+    // this method should not be considered as dead code, see Serializable contract
+  }
+
+  @SuppressWarnings("unused")
+  private int unusedPrivateMethod() { // Noncompliant {{Remove this unused private "unusedPrivateMethod" method.}}
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/UnusedPrivateMethod.java]: According to the specifications of the rules ([RSPEC-3165](https://jira.sonarsource.com/browse/RSPEC-3165)), annotated methods should be ignored.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java (hunk 3 above)
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: @m-g-sonar thank you, missed that.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java (hunk 2 above)
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: @m-g-sonar thank you, missed that.

File: java-checks/src/test/files/checks/UnusedPrivateMethod.java (hunk 4 above)
Comment (mpaladin) [inline java-checks/src/test/files/checks/UnusedPrivateMethod.java]: @m-g-sonar this was not in SONARJAVA-958, I added SONARJAVA-1179 to the current PR which takes care of that.

**Pull request merged**